### PR TITLE
Documentation: Production readiness checklist

### PR DESCRIPTION
**Pull request by StefanRRichter**
Documentation: Production readiness checklist

### CODE REVIEW

File: docs/ops/production_ready.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+you want your job to be (re-)scalable and to have a good upgrade story for your job and new Flink versions.
+
+In the following, we present a collection of configuration options that you should check before your job goes into production.
+
+### Set maximum parallelism for operators explicitly
+
+Maximum parallelism is a configuration parameter that is newly introduced in Flink 1.2 and has important implications
+for the (re-)scalability of your Flink job. This parameter, which can be set on a per-job and/or per-operator granularity,
+determines the maximum parallelism to which you can scale operators. It is important to understand that (as of now) there
+is **now way to increase** this parameter after your job was initially started, except for restarting your job completely
```
Comment (alpinegizmo) [inline docs/ops/production_ready.md]: no way

after your job has been started, except

File: docs/ops/production_ready.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+Maximum parallelism is a configuration parameter that is newly introduced in Flink 1.2 and has important implications
+for the (re-)scalability of your Flink job. This parameter, which can be set on a per-job and/or per-operator granularity,
+determines the maximum parallelism to which you can scale operators. It is important to understand that (as of now) there
+is **now way to increase** this parameter after your job was initially started, except for restarting your job completely 
+from scratch (i.e. with a new state, and not from a previous checkpoint/savepoint). Even if Flink would provide some way
+to change maximum parallelism for existing savepoints in the future, you can already assume that for large states this is 
+likely a long running operation that you want to avoid. At this point, you might wonder why not just to use a very high
+value as default for this parameter. The reason behind this is that high maximum parallelism can have some impact on your
+applications performance and even state sizes, because Flink has to maintain certain meta data for it's ability to rescale which
```
Comment (alpinegizmo) [inline docs/ops/production_ready.md]: application's

metadata for its ability

File: docs/ops/production_ready.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+can increase with the maximum parallelism. In general, you should chose a max parallelism that is high enough to fit your
```
Comment (alpinegizmo) [inline docs/ops/production_ready.md]: choose

File: docs/ops/production_ready.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+- `128` : for all parallelism <= 128.
+- `MIN(nextPowerOfTwo(parallelism + (parallelism / 2)), 2^15)` : for all parallelism > 128.
+
+### Set UUIDs for operators
+
+As mentioned in the documentation for [savepoints]({{ site.baseurl }}/setup/savepoints.html, users should set uids for
+operators. Those operator uids are important for Flink's mapping of operator states to operators which, in turn, is 
+essential for savepoints. By default operator uids are generated by traversing the JobGraph and hashing certain operator 
+properties. While this is comfortable from a user perspective, it is also very fragile to changes on the JobGraph (e.g.
+if you want to exchange an operator). To establish a stable mapping, we need stable operator uids provided by the user
```
Comment (alpinegizmo) [inline docs/ops/production_ready.md]: While this is comfortable from a user perspective, it is also very fragile, as changes to the JobGraph (e.g.
exchanging an operator) will result in new UUIDs.

File: docs/ops/production_ready.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+through `setUid(String uid)`.
+
+### Choice of state backend
+
+Currently, Flink has the limitation that it can only restore the state from a savepoint for the same state backend that
+took the savepoint. For example, this means that we can not take a savepoint with a memory state backend, then change
+the job to use RocksDB state backend and restore. While we are planning to make backends interoperable in the near
```
Comment (alpinegizmo) [inline docs/ops/production_ready.md]: to use a RocksDB

File: docs/ops/production_ready.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+future, they are not yet. This means you should carefully consider which backend you use for your job before going to
+production.
+
+In general, we recommend using RocksDB because this is currently the only state backend that supports large states (i.e.
+state that exceeds the available main memory) and asynchronous snapshots. From our experience, asynchronous snapshots are
+very important for large states because they do not block the operators and Flink can write the snapshots without stopping
+on the stream processing. However, RocksDB can have worse performance that e.g. the memory based state backends. If
```
Comment (alpinegizmo) [inline docs/ops/production_ready.md]: without stopping stream processing

worse performance than, for example, the memory-based state backends

**Comment by uce** [comment]
Looks good to me. Going to merge this...

**Pull request closed without merging**